### PR TITLE
Provide non-atomic versions of many reference counting operations.

### PR DESCRIPTION
**Pull request by swiftix**
#### What's in this pull request?

Provide non-atomic versions of many reference counting operations.

Non-atomic reference counting operations are much faster than their atomic counterparts.

These new entry points will be used by optimizations, which would replace atomic reference counting operations by non-atomic reference counting operations when it is safe to do so. 
#### Resolved bug number: ([SR-248](https://bugs.swift.org/browse/SR-248))

---

Before merging this pull request to apple/swift repository:
- [x] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by swiftix** [comment]
@jckarter Joe, do you mind taking a look?

File: stdlib/public/SwiftShims/RefCount.h [hunk 1]
```diff
@@ -128,6 +140,25 @@ class StrongRefCount {
...
+    while (true) {
+      // If the flag is already set, just fail.
+      if (oldval & RC_PINNED_FLAG) {
+        return false;
+      }
+
+      // Try to simultaneously set the flag and increment the reference count.
+      uint32_t newval = oldval + (RC_PINNED_FLAG + RC_ONE);
+      // Use store __ATOMIC_RELAXED here?
+      if (refCount != oldval)
```
Comment (jckarter) [inline stdlib/public/SwiftShims/RefCount.h]: We don't need a CAS loop here as far as I can see, since the operation is nonatomic. We should be able to update the flags and store.

File: stdlib/public/SwiftShims/RefCount.h [hunk 2]
```diff
@@ -239,6 +282,48 @@ class StrongRefCount {
...
+    // more efficient mask: the check just becomes "is newval nonzero".
+    if ((newval & (RC_COUNT_MASK | RC_PINNED_FLAG | RC_DEALLOCATING_FLAG))
+          != 0) {
+      // Refcount is not zero. We definitely do not need to deallocate.
+      return false;
+    }
+
+    // Refcount is now 0 and is not already deallocating.  Try to set
+    // the deallocating flag.  This must be atomic because it can race
+    // with weak retains.
```
Comment (jckarter) [inline stdlib/public/SwiftShims/RefCount.h]: Would we use a nonatomic release in situations where there may be concurrent weak references?

Comment (gparker42) [inline stdlib/public/SwiftShims/RefCount.h]: You had better not. You can't possibly avoid races against concurrent strengthening of the weak reference (unless you somehow know that all of the weak variables themselves are accessible only from this thread).

File: stdlib/public/SwiftShims/RefCount.h [hunk 3]
```diff
@@ -273,6 +358,47 @@ class StrongRefCount {
...
```
Comment (jckarter) [inline stdlib/public/SwiftShims/RefCount.h]: Same here, the synchronization isn't necessary for a nonatomic operation.

File: stdlib/public/SwiftShims/RefCount.h [hunk 4]
```diff
@@ -239,6 +282,48 @@ class StrongRefCount {
...
+    //
+    // This also performs the before-deinit acquire barrier if we set the flag.
+    static_assert(RC_FLAGS_COUNT == 2,
+                  "fix decrementShouldDeallocate() if you add more flags");
+    uint32_t oldval = 0;
+    newval = RC_DEALLOCATING_FLAG;
+    if (refCount != oldval)
+      return false;
+    __atomic_store_n(&refCount, newval, __ATOMIC_RELAXED);
```
Comment (gparker42) [inline stdlib/public/SwiftShims/RefCount.h]: The lack of barrier here means that your atomic operations are limited to objects that have *never* been used by any other thread, unless you can prove that something else provided the necessary barrier between the other threads' release calls and the start of dealloc.

Comment (jckarter) [inline stdlib/public/SwiftShims/RefCount.h]: @gparker42 Would the release barrier in a `release` that brought the refcount back down to one, or in an `isUniquelyReferenced` query that checks for it, be sufficient? It seems to me we would need one or the other to ever start using nonatomic operations on an unknown object.

Comment (swiftix) [inline stdlib/public/SwiftShims/RefCount.h]: Let's say we can prove that we called "make_unique" on such an object before, e.g. we have a COW data type like Array, Dictionary or Set and we prove that before the call of this operation there was a call that made sure that the buffers used by the COW objects internally are unique. Is it enough to make sure that something else provided the necessary barrier between threads?

Or should we better insert a barrier here to make it more general and to be on the safe side?

Comment (swiftix) [inline stdlib/public/SwiftShims/RefCount.h]: @gparker42 Yes, my question is essentially the same as the one asked by @jckarter. To ever start using the nonatomic operations on an object, we need to be sure that a release barrier was performed by one of the preceding operations, e.g. by a `release` or in most cases by a `isUniquelyReferenced ` operation.

**Comment by swiftix** [comment]
@jckarter @gparker42 I pushed a new version. It addresses @jckarter's comments. 

I have not introduced barriers, because IMHO, the non-atomic APIs are not supposed to be used if the compiler cannot prove that the object is owned by a single thread. And if this can be proven, then there is no need for such barriers, right?

File: stdlib/public/SwiftShims/RefCount.h [hunk 5]
```diff
@@ -273,6 +349,44 @@ class StrongRefCount {
...
   }
+
+  template <bool ClearPinnedFlag>
+  bool doDecrementShouldDeallocateNNonAtomic(uint32_t n) {
+    // If we're being asked to clear the pinned flag, we can assume
+    // it's already set.
+    uint32_t delta = (n << RC_FLAGS_COUNT) + (ClearPinnedFlag ? RC_PINNED_FLAG : 0);
+    uint32_t val = __atomic_load_n(&refCount, __ATOMIC_RELAXED);
+    val -= delta;
+    __atomic_store_n(&refCount, val, __ATOMIC_RELAXED);
```
Comment (jckarter) [inline stdlib/public/SwiftShims/RefCount.h]: I don't think this intermediate store and later assertion really helps us here, since without a barrier we aren't guaranteed to interact with updates from other threads anyway.

File: stdlib/public/SwiftShims/RefCount.h [hunk 6]
```diff
@@ -239,6 +276,45 @@ class StrongRefCount {
...
 
   template <bool ClearPinnedFlag>
+  bool doDecrementShouldDeallocateNonAtomic() {
+    // If we're being asked to clear the pinned flag, we can assume
+    // it's already set.
+    constexpr uint32_t quantum =
+      (ClearPinnedFlag ? RC_ONE + RC_PINNED_FLAG : RC_ONE);
+    uint32_t val = __atomic_load_n(&refCount, __ATOMIC_RELAXED);
+    val -= quantum;
+    __atomic_store_n(&refCount, val, __ATOMIC_RELAXED);
```
Comment (jckarter) [inline stdlib/public/SwiftShims/RefCount.h]: I don't think this intermediate store and later assertion really helps us here, since without a barrier we aren't guaranteed to interact with updates from other threads anyway.

**Comment by jckarter** [comment]
There's still an extra unnecessary store in your `doDecrementShouldDeallocateNonAtomic` implementations, but other than that it LGTM. I'd still wait for @gparker42's approval too.

**Comment by swiftix** [comment]
@gparker42 Greg, could you reply to Joe's and my comments and questions? Do you still think we need barriers?

**Comment by swiftix** [comment]
@jckarter I just spoke with @gparker42 and we agreed that it is safer to have the barriers on the deallocation pass.

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swiftix @gparker42 SGTM. For my own education, why is the barrier necessary if you've already done an `isUniquelyReferenced` check?

**Comment by atrick** [comment]
@gparker42 
As I mentioned to @swiftix, the memory barrier on the refcount decrement is still required for the same reason that it was needed before.

T1: store to x, release x
T2: checkuniq x, release x, dealloc x (release barrier)

Wither a barrier on the normal release, T1 could issue the store after the object is deallocated by T2.

**Comment by jckarter** [comment]
I see, thanks @atrick.

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by atrick** [comment]
@jckarter @gparker42 
For the record, the example I showed is bogus because the reference would not be unique in T1, so its release would be still be atomic. To expose a problem, the same object needs to be unique in one thread, handed off to another thread (written to a global), then accessed as unique in a second thread. Through ARC magic, it might be possible in theory to store the reference to a global without touching the refcount (it's copied to the global as a 'take').

At any rate, proving safety or lack thereof is tricky. I don't anticipate the release barrier to have any measureable performance cost (although @swiftix plans to verify that).

**Comment by swiftix** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@atrick Even then, since "store to x" is nonatomic, and the barrier in "release" isn't a full store fence, couldn't the nonatomic store still be reordered?

**Comment by atrick** [comment]
@jckarter The "release" barrier prevents the current store (in this case, the refcount decrement) from completing prior to any other loads and stores from that thread that appear before it in program order. For example, you could initialize an object, publish the object pointer via a "release" and be sure that other threads only see initialized memory.

Additionally, a "release" barrier cannot be reordered with subsequent barriers, but can be reordered with subsequent nonatomic loads and stores.

**Comment by jckarter** [comment]
> For example, you could initialize an object, publish the object pointer via a "release" and be sure that other threads only see initialized memory.

Don't the other threads still need either a `consume` dependency on the pointer, or an `acquire` barrier, to see the initialization?

**Comment by atrick** [comment]
Other threads would only need an acquire barrier if they were entering a critical region. In my simple object initialization example, only the initializing threads needs any kind of barrier because the object isn't shared across threads until it is "published".

**Comment by gparker42** [comment]
The barriers in retain and release are not intended to provide sufficient synchronization to publish an object to other threads. How could they? In the ordinary simple case you need to store to the global before calling release. Any barrier in release is too late to help your thread synchronization.

There is only one happens-before relation in retain/release: every operation by a thread before that thread called release happens-before the object's deinit begins. Any other synchronization must occur by other means. 

(There might be another happens-before relation with weak variables becoming zero but I think that one would be enforced outside retain/release itself.)

> To expose a problem, the same object needs to be unique in one thread, handed off to another thread (written to a global), then accessed as unique in a second thread.

This is already a race-prone operation. (For example on some architectures there is no guarantee that the second thread sees an initialized object.) Resolution of the other races would almost certainly have enough barriers to fix the happens-before-dealloc problem as well.

**Comment by atrick** [comment]
Sorry for causing confusion with my misleading example. Backing up to the beginning:

My argument is that the nonatomic variants of retain/release should provide the same level of synchronization as the atomic variants, mainly because synchronization is not the thing that the new runtime calls are attempting to optimize, as I mentioned here:
https://bugs.swift.org/browse/SR-249

As @gparker42 just explained, retain/release operations provide almost nothing in the way of synchronization. But there is a guarantee that a thread's own operations complete before a release. Generally, you don't get any happens-before relation across threads without both a release and an acquire (as @jckarter said). The current retain/release synchronization only needs an acquire on the deallocation path because we only care that access to the object happens before it is deallocated.

Now I'm forced to explain my previous comment even though it is irrelevant to this problem:
You can cheat on synchronization by knowing that an object is uniquely referenced. In that case, you could rely on a single hardware release barrier on the store that "publishes" the object pointer to other threads. That's sufficient because we simply don't care about the other thread's order of operations as long as data flow dependencies are respected. But...
(a) you need a release barrier everywhere the pointer may be published
(b) you still need a software memory barrier in the "client" thread so the compiler can't in theory do something like value prediction when it loads the shared reference

Now for what *is* important to this problem: the retain/release synchronization is only relevant to the new runtime calls when this sequence of events occurs...
(1) the object is unique in thread1
(2) the object is shared with thread2
(3) the object is made unique in thread2
(4) the object is deallocated

Step (2) is either a race condition, or is synchronized such that all access to the object in thread1 happens before deallocation.

In conclusion: we don't need the release barrier on the fast decrement path of the new runtime calls for the reason that I thought, but I still think we should keep the barrier because...

- The release barrier should have no performance penalty (and will be verified).

- Race conditions happen, and I would rather not make them harder to debug.

- Our uniqueness check only applies to the strong refcount, and reasoning about synchronization in the presense of weak references is already complicated enough.
